Ctrl shift alt spacemouse

galou opened the pull request:
I implemented Ctrl, Shift, Alt, and Esc support in FreeCAD.

The keys will loose their special function if they are remapped in the configuration.

This is a draft for now because
- testers are needed, I just have a SpacePilot and a SpaceMouse Wireless (which doesn't have the special buttons). Of course people are welcome to offer me their device for testing ;)
- their should be a text in the customization page explaining that the special keys will loose their special feature if remapped. Actually a text is missing explaining how to configure space mouse buttons, even if it's quite, but not completely, intuitive.
- the Esc key is able to close an open sketch but not the "Clarify Selection" context menu.

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->

## Before and After Images
<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->



<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

Commit:
Replace an enum with a enum class

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

Commit:
Modernize and improve C++ style

- Use `[[maybe_unused]]` attribute for unused parameters.
- Use `std::array` for fixed-size arrays.
- Use `const` where applicable.
- Use `std::swap` for swapping values.
- Improve code readability by simplifying string comparisons.
- Define member variables with default initializers.
- Remove unnecessary `.c_str()` calls.
- Declare loop variables locally.
- Return an `std::tuple` rather than using output parameters.
- Calls static methods directly via the class name.

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

Commit:
Return earlier and remove some else branches

Remove some else branches where the if branch ends with a return statement.

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

Commit:
Add support for special spaceball keys

- Add support for mapping spaceball buttons to Ctrl, Shift, Alt, and Esc keys.
- The Esc key works to exit sketch-edit mode for example but not to exit
  the "Clarify Selection" dialog.
- The special keys will loose their special-character if remapped.

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

Commit:
Remove a warning

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

galou commented:
I'm compiling with Qt5 and there's no issue with it. I see though that CI fails because of Qt6. Can someone help with compilation with Qt6?

Commit:
Fix compilation with Qt6

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

pjcreath commented:
There's an awful lot of unnecessary churn.  I think you should submit a PR strictly related to b0f1765.

Also, I'm not entirely sure what functionality you're that commit is adding:  it looks like only Esc and Ctrl are actually handled?  And what are you intending with your changes to `GUIApplication::notify`?

Commit:
Fix compilation on Mac

Commit e094940f79a8 introduced a compilation error on Mac because
of declarations under a `#define` conditional.

Signed-off-by: Gaël Écorchard <gael@km-robotics.cz>

galou commented:
@sw1nn I saw that you're also working on space mouse integration. Could you please test this PR? Thanks.

maxwxyz commented:
@galou could you resolve the conflicts?

maxwxyz commented:
@galou friendly reminder

luzpaz commented:
@galou ping

galou commented:
Thanks for the reminders. I know about it but I'm kinda fighting to find some free time last months.
